src/application/page1/ReservationUnitModal.tsx

```diff
@@ -125,63 +132,156 @@ const Text = styled.span`
   const [selected, setSelected] = useState<ReservationUnit[]>([]);
+  const [purpose, setPurpose] = useState<OptionType | undefined>(undefined);
+  const [reservationUnitType, setReservationUnitType] = useState<
+    OptionType | undefined
+  >(undefined);
+  const [results, setRes] = useState<ReservationUnit[] | [] | undefined>(
+    undefined
+  );
+  const [searching, setSearching] = useState<boolean>(false);
+
+  const purposeOptions = [emptyOption].concat(options.purposeOptions);
+  const reservationUnitTypeOptions = [emptyOption].concat(
+    options.reservationUnitTypeOptions
+  );
 
   const { t } = useTranslation();
 
-  const results = useAsync(async () => {
-    if (q === null) {
-      return [];
-    }
-    return getReservationUnits({ search: q });
-  }, [q]);
+  const searchResults = async () => {
+    setSearching(true);
+    const searchCriteria = {
+      ...(searchTerm && { search: searchTerm }),
+      ...(purpose && { purpose: purpose.value }),
+      ...(reservationUnitType && {
+        reservationUnitType: reservationUnitType.value,
+      }),
+    };
+
+    const reservationUnits = await getReservationUnits(searchCriteria);
+    setRes(reservationUnits);
+    setSearching(false);
+  };
+
+  if (results === undefined && searching === false) searchResults();
+  const emptyResult = results?.length === 0 && (
+    <div>{t('common.noResults')}</div>
+  );
 
   const filtered = currentReservationUnits.concat(selected).map((ru) => ru.id);
 
   return (
     <MainContainer>
       <Heading>{t('ReservationUnitModal.heading')}</Heading>
       <Text>{applicationPeriod.name}</Text>
-      <StyledSearchInput
-        label={t('ReservationUnitModal.searchTermLabel')}
-        onSubmit={(e) => {
-          setQ(e);
-        }}
-      />
+      <Filters>
+        <TextInput
+          id="reservationUnitSearch.search"
+          label={t('ReservationUnitModal.searchTermLabel')}
+          onChange={(e: ChangeEvent<HTMLInputElement>): void => {
+            setSearchTerm(e.target.value);
+          }}
+        />
+        <Select
+          id="reservationUnitSearch.purpose"
+          placeholder={t('common.select')}
+          options={purposeOptions}
+          label={t('ReservationUnitModal.searchPurposeLabel')}
+          onChange={(selection: OptionType): void => {
+            setPurpose(selection);
+          }}
+          defaultValue={emptyOption}
+        />
+        <Select
+          id="reservationUnitSearch.reservationUnitType"
+          placeholder={t('common.select')}
+          options={reservationUnitTypeOptions}
+          label={t('ReservationUnitModal.searchReservationUnitTypeLabel')}
+          onChange={(selection: OptionType): void => {
+            setReservationUnitType(selection);
+          }}
+          defaultValue={emptyOption}
+        />
+      </Filters>
+      <ButtonContainer>
+        <SearchButton onClick={() => searchResults()}>
```

Comment:
Looks like this button now tries to submit the form behind the modal tool, need to add call to "e.preventDefault()' ?

Reply:
Yes, good catch. Fixed.